src/syntax/ppx_sedlex.ml

```diff
@@ -253,24 +264,195 @@ let gen_recflag auto =
     Nonrecursive
   with Exit -> Recursive
 
+let gen_cset ~loc x char_set =
+  let interval a b =
+    [%expr
+      [%e eint ~loc a] <= [%e evar ~loc x]
+      && [%e evar ~loc x] <= [%e eint ~loc b]]
+  in
+  match char_set with
+    | (a, b) :: l ->
+        List.fold_left
+          (fun acc (a, b) -> [%expr [%e acc] || [%e interval a b]])
+          (interval a b) l
+    | [] -> assert false
+
+let gen_trace lexbuf traces i = function
+  | (_, []), _ -> []
+  | (_, aliases), _ ->
+      let loc = default_loc in
+      let initial, trans, finals = traces.(i) in
+      let alias_indexes =
+        List.to_seq aliases
+        |> Seq.mapi (fun i { txt = alias } -> (alias, i))
+        |> StringMap.of_seq
+      in
+      let find_index alias = StringMap.find alias alias_indexes in
+      let result_arrays =
+        [
+          value_binding ~loc
+            ~pat:[%pat? __sedlex_aliases_starts]
+            ~expr:(pexp_array ~loc (List.map (fun _ -> [%expr 0]) aliases));
+          value_binding ~loc
+            ~pat:[%pat? __sedlex_aliases_stops]
+            ~expr:(pexp_array ~loc (List.map (fun _ -> [%expr 0]) aliases));
+        ]
+      in
+      let aux_fun =
+        let gen_start alias e =
+          let alias_index = find_index alias in
+          [%expr
+            __sedlex_aliases_starts.([%e eint ~loc alias_index]) <- __sedlex_pos;
+            [%e e]]
+        in
+        let gen_stop alias e =
+          let alias_index = find_index alias in
+          [%expr
+            __sedlex_aliases_stops.([%e eint ~loc alias_index]) <- __sedlex_pos;
+            [%e e]]
+        in
+        let unreachable_case =
+          case ~lhs:[%pat? _] ~guard:None ~rhs:[%expr assert false]
+        in
+        let trans_cases =
+          List.map
+            (fun {
+                   Sedlex.curr_state;
+                   curr_node;
+                   prev_state;
+                   prev_node;
+                   char_set;
+                   starts;
+                   stops;
+                 } ->
+              let lhs =
+                ppat_tuple ~loc
+                  [
+                    pint ~loc curr_state;
+                    pint ~loc curr_node;
+                    pint ~loc prev_state;
+                  ]
+              in
+              let guard = Some (gen_cset ~loc "__sedlex_code" char_set) in
```

Comment:
IIUC, the guard is only useful when that are duplicate triples `(curr_state, curr_node, prev_state)` in `trans`

Reply:
Yeah. My mistake

Reply:
Based on the fix you've committed, I think you misunderstand my point.

The generated code look like 
```ocaml
           | (4, 9, 3) when (98 <= __sedlex_code) && (__sedlex_code <= 98) ->
               (__sedlex_aliases_stops.(1) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 3 7 __sedlex_rest)
           | (0, 5, 0) when (97 <= __sedlex_code) && (__sedlex_code <= 97) ->
               (__sedlex_aliases_stops.(0) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 0 3 __sedlex_rest)
           | (4, 21, 4) when (98 <= __sedlex_code) && (__sedlex_code <= 98)
               ->
               (__sedlex_aliases_stops.(0) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 4 18 __sedlex_rest)
``` 
the `when` guards are only useful when tripes  (e.g. `(4, 9, 3)`) are not unique.

```ocaml
           | (4, 9, 3)  ->
               (__sedlex_aliases_stops.(1) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 3 7 __sedlex_rest)
           | (0, 5, 0)  ->
               (__sedlex_aliases_stops.(0) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 0 3 __sedlex_rest)
           | (4, 21, 4)  ->
               (__sedlex_aliases_stops.(0) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 4 18 __sedlex_rest)
```

Reply:
This now looks good.